bftengine/tests/simpleTest/test_comm_config.cpp

```diff
@@ -0,0 +1,175 @@
+#include <vector>
+#include <fstream>
+
+#include "CommFactory.hpp"
+#include "threshsign/ThresholdSignaturesSchemes.h"
+#include "KeyfileIOUtils.hpp"
+#include "config_file_parser.hpp"
+
+using bftEngine::PlainUdpConfig;
+using bftEngine::PlainTcpConfig;
+using bftEngine::ReplicaConfig;
+using BLS::Relic::BlsThresholdFactory;
+using std::pair;
+using std::map;
+using std::string;
+using std::vector;
+
+#define CLIENTS_CONFIG  "clients_config"
+#define REPLICAS_CONFIG "replicas_config"
+
+const char* TestCommConfig::ip_port_delimiter_ = ":";
+const std::string TestCommConfig::default_ip_ = "127.0.0.1";
+
+//////////////////////////////////////////////////////////////////////////////
+// Create a replica config for the replica with index `replicaId`.
+// inputReplicaKeyfile is used to read the keys for this replica, and default
+// values are loaded for non-cryptographic configuration parameters.
+void TestCommConfig::GetReplicaConfig(uint16_t replica_id,
+                                      bftEngine::ReplicaConfig* out_config) {
+
+  std::string key_file_name = "private_replica_" + std::to_string(replica_id);
+  std::ifstream keyfile(key_file_name);
+  if (!keyfile.is_open()) {
+    throw std::runtime_error("Unable to read replica keyfile.");
+  }
+
+  bool success = inputReplicaKeyfile(keyfile, key_file_name, *out_config);
+  if (!success)
+    throw std::runtime_error("Unable to parse replica keyfile.");
+
+  // Set non-cryptographic configuration
+  out_config -> statusReportTimerMillisec = 2000;
+  out_config -> concurrencyLevel = 1;
+}
+
+std::unordered_map <NodeNum, NodeInfo> TestCommConfig::SetUpConfiguredNodes(
+    bool is_replica, const std::string& config_file_name, uint16_t id,
+    std::string& ip, uint16_t& port, uint16_t& num_of_clients,
+    uint16_t& num_of_replicas) {
+
+  ConfigFileParser config_file_parser(logger_, config_file_name);
+  if (!config_file_parser.Parse()) {
+    LOG_FATAL(logger_, "Failed to parse configuration file: " <<
+                       config_file_name);
+    exit(-1);
+  }
+  num_of_clients  = static_cast<uint16_t>(
+      config_file_parser.Count(CLIENTS_CONFIG));
+  num_of_replicas = static_cast<uint16_t>(
+      config_file_parser.Count(REPLICAS_CONFIG));
+  if ((is_replica && (id + 1 > num_of_replicas)) ||
+      (!is_replica && (id + 1 > num_of_clients + num_of_replicas))) {
+    LOG_FATAL(logger_, "Wrong number of clients/replicas configured: " <<
+                       "numOfClients=" << num_of_clients <<
+                       ", numOfReplicas=" << num_of_replicas);
+    exit(-1);
+  }
+  vector<string> replicas = config_file_parser.GetValues(REPLICAS_CONFIG);
+  vector<string> clients  = config_file_parser.GetValues(CLIENTS_CONFIG);
+  std::unordered_map <NodeNum, NodeInfo> nodes;
+  int k = 0;
+  for (int i = 0; i < (num_of_replicas + num_of_clients); i++) {
+    vector<string>& current_vector = replicas;
+    if ((k == i) && (i >= num_of_replicas)) { // Ensure this happens only once.
```

Comment:
Could you add a little more detail about *why* this needs to happen only once, please? If I understand correctly, it's because `k` is used to map the principle ID, `i`, to the replica/client ID.

Would it be cleaner to just say `k = (i < num_replicas) ? i : i - num_replicas`, instead of this stateful variable? That would also get rid of the `k++` / `k - 1` juggle below.

Reply:
I'll change the comment, np.
Regarding the code: I still need to switch between replica and client vectors, so "if" should remain. Besides that, after a switch to clients vector, k should be reset to 0 and it will not be a function of "i", so I prefer to leave this code as is.